### PR TITLE
Implement log system

### PR DESCRIPTION
**Pull request by OJarrisonn**
This PR ain't complete so **DO NOT MERGE IT YET**, this is just to discuss around the current status of this feature.

- Introduces a new dependency: [chrono](https://crates.io/crates/chrono)
- Introduces a new configuration option: `log_path`
- Adds a `Logger` singleton that is non-thread safe
- Known issue: the user must manually create `log_path` folder before using
- Known issue: memory utilization: log entries are stored both in memory and in disk
- No logging level (INFO, WARN, FAIL), etc

This PR also attends to the new need to document code, so most of the new features are well-documented

Closes: #36

### CODE REVIEW

**Comment by davidbtadokoro** [comment]
Hi @OJarrisonn, thanks a lot for this PR! I will take a closer look at it to begin discussions ASAP.

**Review by davidbtadokoro — CHANGES_REQUESTED** [review]
Hi @OJarrisonn, and thank you so much for this PR! Sorry for the delay, but I wanted to give it the attention it deserves.

## Comments about the PR

Before bringing my thoughts and starting the "high-level" discussion, I want to address small things about the PR itself, which I think you didn't bother with because of the "begin discussion" idea of the PR. I am just laying them here for the record and for when we move on to refine and merge it.

1. You can squash all commits into one
2. When reviewing and testing the change, I produced feedback that will be important to refinement for when we merge it, so I left some inline review comments
3. Always add your signed-off by

## Discussion about the feature itself

I've tested a handful of scenarios and, overall, the feature works as intended :+1: :+1:

IMHO, we are okay in robustness as we persist logs on the fly and keep a separate data structure for printing (the change to the panic hook also helps avoid losing logs). The restriction on single-threaded isn't an issue for the moment, so I vouch for crossing this bridge when we need :sweat_smile: 

At first, I tried adapting your implementation to remove the buffer with the vector of strings. As we are persisting the logs immediately, I thought it was redundant to keep an identical copy in memory instead of just printing the contents of the log file when flushing. However, I gave up on this idea because we probably won't want to print every log to the terminal at the end of execution. On the other hand, there will be some warnings and errors that we want the user to be aware of, so I am leaning towards having (1) the log file that will amass **all** events that are logged and (2) a buffer for the "logs to print". In the future, it wouldn't be hard to add a "verbose" flag/config to patch-hub to print all logs to the terminal, in which case we would fall in the current implementation, i.e., the buffer to print and file contents will be the same.

Below is a draft of change on `write` for logging with two levels (to print, or not) in which I just add a boolean, essentially. 

```rust
    pub fn write(&mut self, msg: &str, print: bool) {
        let msg = format!(
            "[{}] {}",
            chrono::Local::now().format("%Y-%m-%d %H:%M:%S"),
            msg
        );

        let log_file = self.log_file
            .as_mut()
            .expect("Log file not initialized, make sure to call Logger::init_log_file() before writing to the log file");
        writeln!(log_file, "{msg}")
            .expect("Failed to write to log file");

        if print {
            self.logs_to_print.push(msg);
        }
    }
```

Obs.: I've renamed `buffer` to `logs_to_print`.

## Conclusion

I think we are on the right track and that this isn't so raw as you've said :rofl: 

Anyway, hit me with your thoughts, and remember that I am open to discussing any points I've brought up.

Great work!!!

File: src/app/config.rs [hunk 1]
```diff
@@ -6,6 +6,10 @@ pub struct Config {
     pub bookmarked_patchsets_path: String,
     pub mailing_lists_path: String,
     pub reviewed_patchsets_path: String,
+    /// The path to the logs directory
+    /// 
+    /// A folder inside the `patch-hub/logs` directory
```
[the remark this answers is not in the document]
Comment (OJarrisonn) [reply src/app/config.rs:11]: I like more verbose documentation but i applied your suggestion

File: src/app/logging.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    /// # Safety
+    ///
+    /// It's safe to use in single-threaded scenarios only
+    ///
+    /// # Examples
+    /// ```rust norun
+    /// // Get the logger singleton
+    /// Logger::logger().init_log_file(&config); // Initialize the log file
+    /// Logger::logger().write("This is a log message"); // Write a message to the log file
+    /// ```
```
[the remark this answers is not in the document]
Comment (OJarrisonn) [reply src/app/logging.rs:51]: We do have examples in the other functions, but i think it's worth to show the usage of a function in it's documentation.

File: src/app/logging.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    /// # Examples
+    /// ```rust norun
+    /// // Once you get the config struct...
+    /// let config = Config::build();
+    /// // ... initialize the log file
+    /// Logger::logger().init_log_file(&config);
+    /// ```
+    pub fn init_log_file(&mut self, config: &Config) {
+        if let None = self.log_file {
+            let path = config.logs_path.clone();
```
[the remark this answers is not in the document]
Comment (OJarrisonn) [reply src/app/logging.rs:138]: Also, no need to clone it

File: src/app.rs [hunk 4]
```diff
@@ -369,6 +371,10 @@ impl App {
             Err(_) => reviewed_patchsets = HashMap::new(),
         }
 
+        // Initialize the logger before the app starts
+        Logger::logger().init_log_file(&config);
+        Logger::logger().write("Patch-hub started");
```
[the remark this answers is not in the document]
Comment (OJarrisonn) [reply src/app.rs:376]: Sure thing

**Comment by davidbtadokoro** [comment]
Hi @OJarrisonn, and thank you so much for the update! Also, marvelous work on the logging levels :+1: 

I've made changes to your PR, which you can check at [this branch](https://github.com/kworkflow/patch-hub/tree/OJarrisonn-logging-reviewed). Note that the commit messages contain a `Maintainer edits` section that overviews my modifications.

 If you agree with this proposal, we can move to merging!

**Comment by OJarrisonn** [comment]
I just disagree with the changes in the fields names for `Logger`, i rather use shorter names like: `file`, `filepath`, `logs` and `level`

Also, was the log level filtering logic really inverted? `level <= self.print_level` means that if we set the filter level to `Warn` we will only print `Info` and `Warn` logs, and discard `Error`. Is this really the expected behaviour?

**Comment by davidbtadokoro** [comment]
> I just disagree with the changes in the fields names for `Logger`, i rather use shorter names like: `file`, `filepath`, `logs` and `level`

I agree that whenever possible, we should favor shorter names. By now, you've gathered that I am really verbose in names :sweat_smile: , but at this time, I think `logs` and `level` is too ambiguous. Imagine someone who first sees the code; the name `logs` would prompt him to think that this vector will amass all logs, while this isn't true; also, `level` doesn't convey its means.

> Also, was the log level filtering logic really inverted? `level <= self.print_level` means that if we set the filter level to `Warn` we will only print `Info` and `Warn` logs, and discard `Error`. Is this really the expected behaviour?

Unfortunately, it is indeed inverted... Try adding a `warn()` and an `error()` right below the `info("patch-hub started")` and changing the print level of the Logger to `Warning`. Notice that the warning and the error will be printed when the warning and info should be printed instead.

Simulating the code, if the precedence is `Info < Warning < Error`, if `self.print_level == Warning`, the check

```rust
if self.print_level <= Info
``` 

will be false, making this inversion...

**Comment by OJarrisonn** [comment]
Yeah, my approach is to print logs that are more severe than the filter level. We should always print the errors, while Info aint always relevant to the common user. Error messages always are relevant so fhe user might try to fix it by himself

Remember that the users of patch-hub are kernel developers, so we must provide the kind of logging that a developer might expect

In your approach we will always see the "patch-hub started" message when closing the app. Is this messeage really that important that we always print it to the terminal?

**Comment by OJarrisonn** [comment]
Yes

**Comment by davidbtadokoro** [comment]
Nice, going to do this soon.

Just as a head up, I might still have to solve some merge conflicts due to the last merged change. This probably won't change anything "important" or that we discussed, but just to keep the process transparent.

**Comment by davidbtadokoro** [comment]
Hey @OJarrisonn, just merged the change into the unstable branch :+1: 

Thank you so much for your continuous effort and great work! The logging feature looks really neat and will be crucial for users and developers alike :fire:

**Pull request closed without merging**